### PR TITLE
DNM [nrf mergeup] Upmerge 171120

### PR DESCRIPTION
**Pull request by ioannisg**
In progress

pending: re-apply the reverted downstream commits.

### CODE REVIEW

**Comment by ioannisg** [comment]
Comments:
FYI @mbolivar-nordic  @hakonfam @sigvartmh @nvlsianpu @carlescufi 

- 64617d8 was pushed, so we are able to also revert 80eb854, which was not caught by ncs-compare, although a relevant patch was merged upstream.
- 026a4e5, 75fd8f2, 01dd7b3, ac05701, 22373ec are standard git-reverts, so the corresponding reverted commits must be reapplied as is.
- 06d7123 is a **_manual revert_** I did, to be able to achieve a conflict-free upmerge. Without it, I would have to revert, again, the commit by Sebastian B, that introduces PM (along with several other dependencies); this way we minimize the effort in next up-merges. However, the reverts done in  06d7123  must be re-applied.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
>     * I would have to revert, again, the commit by Sebastian B, that introduces PM (along with several other dependencies)

If that's what it takes, that's what has to be done. 

Here is the `west ncs-loot -z 35264cc214 mcuboot` output for this PR right now:

```
- 7718705e8b500aabbfa1cc720f082bfc048bb9f5 [nrf noup] ci: add Jenkins-based CI for NCS
- 228e8868663ce009c6b1d1a18315ef3c5d75b105 [nrf toup] zephyr: add support for Thingy:91
- 45f0833a8dee3549ef661137dd03e54d180ae8d8 [nrf noup] zephyr: Remove duplication from Kconfig and cmake
- 68f2bcc798113500cf4abf50af6e6def60ba384a [nrf noup] boot: bootutil: Add shared crypto functions for ECDSA
- e6e68505af12b6288da5b7215662bd711ec6e9c0 [nrf noup] boot: zephyr: Call fw_info_ext_api_provide() before booting
- 04514d6a2236050ac8abd89d72f59dd175a0265e [nrf noup] treewide: add NCS partition manager support
- 59e02e27dafd41456c54bf45d0b2d95ab15b6e0f [nrf noup] zephyr: lock mcuboot using fprotect before jumping
- f35f763bbe7e698b3c275c548faae3537061e9c7 [nrf noup] boot: zephyr: nrf53 network core bootloader implementation
- 1788ad7a9894780bf27c68d98aa87e833f170c7d [nrf noup] kconfig: use cc310 for ecdsa by default
- 4d1b74b32f7f49692fc40132c7c5e84030ef6a85 [nrf noup] kconfig: enable external crypto config
- 74e2e0f30c4e48738dd5014bd88d31f90d00f9c5 [nrf fromtree] zephyr: Fix serial recovery compilation warnings
- 54238adc9a5912ace04b35c8a642faa1e731fa74 [nrf fromtree] boot: zephyr: reset SPLIM registers before boot
- 45a94cb234ec33a015677f106cbb937b6ff81569 [nrf fromtree] boot: zephyr: Disable HW stack protection
- d0634dd15ca4541a93bc22d9e1aa0fad07e95bc0 [nrf fromtree] doc: fix link to external page in PORTING
- 80eb85475da3d3a8408170523e7cb6d1063a9757 [nrf fromtree] kconfig: zephyr: provide logic ...
- 32640591847ec648fb5b0cee85c9dc57db7cb3bf [nrf fromtree] bootutil: fix swap-move brick with padded image0
- 931d96b455c536fbd33e50b27f89aa384b054590 [nrf fromtree] bootutil: fix boostrapping in swap-move
- 02833faca1a373767c67aed29df5eb9e0cfbc9fd [nrf fromlist] zephyr: update name of nrf5340 dk
- b03e43f31a96619287bbd0b43545152f17eab0b6 [nrf fromlist] zephyr: enable progressive erase on all nRF boards
- a8213955e7113e3dcc23f17318829dc6b695ce67 [nrf fromlist] doc: clean up multi-image documentation
- 85645596570de42cda3cfd60a0cc5d02813947af [nrf fromlist] bootutil: copy image size with fast overwrite
- e9d64d45865fdd4943a01f55c5096643c399d52a [nrf temphack] Kconfig: add accessor for flash write block size
- a8cc9879747019ba0459b900920296365abae4b1 [nrf temphack] do_boot: clean peripherals state before boot
- 4c5ff85bbd6f218147cc4f29641e88957ad09a64 [nrf temphack] boot: bootutil: Word align read buffer in loader
- ac052356a507e309d9da7246f3c16dde4d46aea3 [nrf temphack] Kconfig: select 8-bits flash access for recovery mode
```

And the upstream commit that's being merged in is `5a6e181`. If you try to cherry pick the `ncs-loot` output onto `5a6e181` in order, you get a conflict ... at the partition manager support patch:

```
CONFLICT (content): Merge conflict in boot/zephyr/CMakeLists.txt
error: could not apply 04514d6... [nrf noup] treewide: add NCS partition manager support
```

> this way we minimize the effort in next up-merges. 

The point of revert-and-reapply is not to minimize effort in the next upmerge, it's to make sure that the tree can be rebased into a linear history at any time.

**Comment by ioannisg** [comment]
> If that's what it takes, that's what has to be done.

@mbolivar-nordic I am not sure I fully-agree; I think we "could" try to be a bit smarter here. The optimization problem is
```
minimize the number of reverted commits
subject to: linear history, conflict-free merges, etc.
```

If we are able to push a "noup" commit that makes us revert less stuff in the next upmerge, we could exploit in case-by-case basis. But I don't have a super strong opinion. I also see the value of a standardized procedure.

So I followed the recommendation and reverted all commits that conflict (along with their dependencies).

**Comment by ioannisg** [comment]
@hakonfam , @nvlsianpu all patches starting from ac05235 need to be re-applied, _unless_ you think they are not needed any more. In that case please, add a comment for each one that is left behind.

**Comment by ioannisg** [comment]
I mean, ac05235,  4c5ff85, a8cc987 and onwards.

**Comment by nvlsianpu** [comment]
https://github.com/nrfconnect/sdk-mcuboot/pull/130/commits/9f3d9db74473f94a38e62e43e8253ab25a006d2a should be left behind.
https://github.com/nrfconnect/sdk-mcuboot/pull/130/commits/f48de7a5237886a994984528b1cb1fb595620e1b removes need for it.

File: boot/bootutil/src/loader.c [hunk 1]
```diff
@@ -751,10 +751,11 @@ boot_validated_swap_type(struct boot_loader_state *state,
         reset_addr = vtable[1];
 #ifdef PM_S1_ADDRESS
         const struct flash_area *primary_fa;
-        rc = flash_area_open(flash_area_id_from_multi_image_slot(
+        int rc = flash_area_open(flash_area_id_from_multi_image_slot(
```
Comment (ioannisg) [inline boot/bootutil/src/loader.c:754]: Fine with me. Can yo squash this then?

**Comment by hakonfam** [comment]
> [9f3d9db](https://github.com/nrfconnect/sdk-mcuboot/commit/9f3d9db74473f94a38e62e43e8253ab25a006d2a) should be left behind.
> [f48de7a](https://github.com/nrfconnect/sdk-mcuboot/commit/f48de7a5237886a994984528b1cb1fb595620e1b) removes need for it.
> 
> same [3ab66d8](https://github.com/nrfconnect/sdk-mcuboot/commit/3ab66d826fe5d98a69047ce1f0e0cdd3dff70888) - , can be removed, all flash drivers supports unaligned reads since a while.

This should be fixed now.

**Review by nvlsianpu — APPROVED** [review]
The upmerge process was done properly

**Review by de-nordic — APPROVED** [review]
Looks OK.
Manually tested CONFIG_SINGLE_APPLICATIONS_SLOT and it works OK.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
I'm still seeing conflicts when I try to rebase.

west ncs-loot says:

```
NCS commit (manifest-rev): 4b41ab6ccaee6c54f37d351b51b85f94499459f7, upstream commit: c71d2186077f9fd5f6d1aa53ee3f09dc41ce78f6
OOT patches: 13 total
- 7718705e8b500aabbfa1cc720f082bfc048bb9f5 [nrf noup] ci: add Jenkins-based CI for NCS
- 228e8868663ce009c6b1d1a18315ef3c5d75b105 [nrf toup] zephyr: add support for Thingy:91
- e6e68505af12b6288da5b7215662bd711ec6e9c0 [nrf noup] boot: zephyr: Call fw_info_ext_api_provide() before booting
- 59f2d3ca15ce58d2e5ef22f2db6bec48f82e6a00 [nrf noup]: remove unnecessary default value
- 7888f0ad4df6eb10fd8c732cb84656a4e7030ad1 [nrf noup] zephyr: Remove duplication from Kconfig and cmake
- f20a1a1c10e3949d1d12bd0621518eccb4cd69f2 [nrf noup] boot: bootutil: Add shared crypto for ECDSA and SHA
- 069b603167a1e8f1491e44139dc099c8065fdf28 [nrf noup] treewide: add NCS partition manager support
- c21b25a9a8f60fa2e10c820be0166ac87de52fdf [nrf noup] zephyr: lock mcuboot using fprotect before jumping
- 3d745d389b10beb1e333a8ea65d3590c2f06b965 [nrf temphack] Kconfig: add accessor for flash write block size
- 337357898d46250584c2d400ea3ae57770ae99b8 [nrf noup] boot: zephyr: nrf53 network core bootloader implementation
- 5e0fc4ec914b0765920c00a994f3dd30c93e4136 [nrf noup] kconfig: use cc310 for ecdsa by default
- a2c9384795e897f52f5eee840ac63f04633335f3 [nrf noup] kconfig: enable external crypto config
- 4b41ab6ccaee6c54f37d351b51b85f94499459f7 [nrf temphack] do_boot: clean peripherals state before boot
```

And I get a different conflict when I cherry-pick:

```

CONFLICT (content): Merge conflict in boot/zephyr/Kconfig
error: could not apply 59f2d3c... [nrf noup]: remove unnecessary default value
```

I'll keep looking into this after codeowners meeting, but want to put a -1 on this again until we get the PR rebasable.

**Comment by mbolivar-nordic** [comment]
We can just skip the conflicting patch during the cherry-pick and we get a zero-diff rebase, which I have posted here:

https://github.com/mbolivar-nordic/sdk-mcuboot/tree/upmerge-130-rebase

This has an empty diff with the current tip of this PR, 4b41ab6.

I think we could make an exception and allow this PR to be merged if we make it very clear in the commit log for the conflicting patch that it is a 'temphack' which should be skipped when rebasing. That won't cause difficulties when it is time to rebase for the next NCS release.

**Review by mbolivar-nordic — APPROVED** [review]

**Review by utzig — APPROVED** [review]

**Pull request merged**